File: AlgoExpert/1. Easy/Python/12. Selection Sort/Selection Sort.py
# Best: O(n^2) time | O(1) space
# Average: O(n^2) time | O(1) space
# Worst: O(n^2) time | O(1) space

def selectionSort(array):
	currentIdx = 0
	while currentIdx < len(array) - 1:
		smallestIdx = currentIdx
		for i in range(currentIdx + 1, len(array)):
			if array[smallestIdx] > array[i]:
				smallestIdx = i
		swap(currentIdx, smallestIdx, array)
		currentIdx += 1
	return array

def swap(i, j, array):
	array[i], array[j] = array[j], array[i]

File: AlgoExpert/1. Easy/Java/12. Selection Sort/Selection Sort.java
class Program {

	// Best: O(n^2) time | O(1) space
	// Average: O(n^2) time | O(1) space
	// Worst: O(n^2) time | O(1) space

	public static int[] selectionSort(int[] array) {
		if (array.length == 0) {
			return new int[] {};
		}
		int startIdx = 0;
		while (startIdx < array.length - 1) {
			int smallestIdx = startIdx;
			for (int i = startIdx + 1; i < array.length; i++) {
				if (array[smallestIdx] > array[i]) {
					smallestIdx = i;
				}
			}
			swap(startIdx, smallestIdx, array);
			startIdx++;
		}
	return array;
	}

	public static void swap(int i, int j, int[] array) {
		int temp = array[j];
		array[j] = array[i];
		array[i] = temp;
	}
}